lightning-transaction-sync/src/electrum.rs

```diff
@@ -0,0 +1,446 @@
+						},
+						Err(err) => {
+							// (Semi-)permanent failure, retry later.
+							log_error!(self.logger,
+								"Failed during transaction sync, aborting. Synced so far: {} confirmed, {} unconfirmed.",
+								num_confirmed,
+								num_unconfirmed
+							);
+							sync_state.pending_sync = true;
+							return Err(TxSyncError::from(err));
+						}
+					}
+
+					// Update the best block.
+					for c in &confirmables {
+						c.best_block_updated(&tip_header, tip_height);
+					}
+				}
+
+				match self.get_confirmed_transactions(&sync_state) {
+					Ok(confirmed_txs) => {
+						// Double-check the tip hash. If it changed, a reorg happened since
+						// we started syncing and we need to restart last-minute.
+						if self.check_has_chain_moved(&mut sync_state, &mut tip_header,
+							&mut tip_height)?
+						{
+							continue;
+						}
+
+						num_confirmed += confirmed_txs.len();
+						sync_state.sync_confirmed_transactions(
+							&confirmables,
+							confirmed_txs,
+						);
+					}
+					Err(InternalError::Inconsistency) => {
+						// Immediately restart syncing when we encounter any inconsistencies.
+						log_debug!(self.logger,
+							"Encountered inconsistency during transaction sync, restarting.");
+						sync_state.pending_sync = true;
+						continue;
+					}
+					Err(err) => {
+						// (Semi-)permanent failure, retry later.
+						log_error!(self.logger,
+							"Failed during transaction sync, aborting. Synced so far: {} confirmed, {} unconfirmed.",
+							num_confirmed,
+							num_unconfirmed
+						);
+						sync_state.pending_sync = true;
+						return Err(TxSyncError::from(err));
+					}
+				}
+				sync_state.last_sync_hash = Some(tip_header.block_hash());
+				sync_state.pending_sync = false;
+			}
+		}
+		log_debug!(self.logger,
+			"Finished transaction sync at tip {} in {}ms: {} confirmed, {} unconfirmed.",
+			tip_header.block_hash(), start_time.elapsed().as_millis(), num_confirmed,
+			num_unconfirmed);
+		Ok(())
+	}
+
+	fn check_has_chain_moved(&self, sync_state: &mut SyncState, cur_tip_header: &mut BlockHeader, cur_tip_height: &mut u32) -> Result<bool, InternalError> {
+		let check_notification = self.client.block_headers_subscribe()?;
+		let check_tip_hash = check_notification.header.block_hash();
+
+		// Restart if either the tip changed or we got some divergent tip
+		// change notification since we started. In the latter case we
+		// make sure we clear the queue before continuing.
+		let mut restart_sync = check_tip_hash != cur_tip_header.block_hash();
+		while let Some(queued_notif) = self.client.block_headers_pop()? {
+			if queued_notif.header.block_hash() != check_tip_hash {
+				restart_sync = true
+			}
+		}
+
+		if restart_sync {
+			*cur_tip_header = check_notification.header;
+			*cur_tip_height = check_notification.height as u32;
+			log_debug!(self.logger,
+				"Encountered inconsistency during transaction sync, restarting.");
+			sync_state.pending_sync = true;
```

Comment:
Could we leave this at the call site given all the other modifications of `pending_sync` are there? Would be easier to see where it is modified that way and save needing to pass the `SyncState` parameter here.